lib/onelogin/ruby-saml/response.rb

```diff
@@ -112,78 +151,327 @@ def status_message
         validate_conditions(soft)     &&
+        validate_destination(soft)    &&
+        validate_audience(soft)       &&
         validate_issuer(soft)         &&
-        document.validate_document(get_fingerprint, soft) &&
-        validate_success_status(soft)
+        validate_session_expiration(soft) &&
+        validate_subject_confirmation(soft) &&
+        document.validate_document(get_fingerprint, soft)
+      end
+
+      # Validates that the SAML Response provided in the initialization is not empty, 
+      # also check that the setting and the IdP cert were also provided 
+      # If fails, the error is added to the errors array.
+      # @param soft [Boolean] soft Enable or Disable the soft mode (In order to raise exceptions when the response is invalid or not)
+      # @return [Boolean] True if the required info is found, otherwise False if soft=True
+      # @raise [ValidationError] if soft == false and validation fails
+      #
+      def validate_response_state(soft = true)
+        if response.nil? or response.empty?
+          @errors << "Blank SAML Response"
+          return soft ? false : validation_error("Blank SAML Response")
+        end
+
+        if settings.nil?
+          @errors << "No settings on SAML Response"
+          return soft ? false : validation_error("No settings on SAML Response")
+        end
+
+        if settings.idp_cert_fingerprint.nil? && settings.idp_cert.nil?
+          @errors << "No fingerprint or certificate on settings"
+          return soft ? false : validation_error("No fingerprint or certificate on settings")
+        end
+
+        true
+      end
+
+      # Validates that the SAML Response contains an ID 
+      # If fails, the error is added to the errors array.
+      # @return [Boolean] True if the SAML Response contains an ID, otherwise returns False
+      #
+      def validate_id()
+        unless id(document)
+          @errors << "Missing ID attribute on SAML Response"
+          return false
+        end
+        true
+      end
+
+      # Validates the SAML version (2.0)
+      # If fails, the error is added to the errors array.
+      # @return [Boolean] True if the SAML Response is 2.0, otherwise returns False
+      #
+      def validate_version()
+        unless version(document) == "2.0"
+          @errors << "Unsupported SAML version"
+          return false
+        end
+        true
       end
 
+      # Validates that the SAML Response only contains a single Assertion (encrypted or not).
+      # If fails, the error is added to the errors array.
+      # @return [Boolean] True if the SAML Response contains one unique Assertion, otherwise False
+      #
+      def validate_num_assertion()
+        assertions = REXML::XPath.match(document, "//a:Assertion", { "a" => ASSERTION })
+        encrypted_assertions = REXML::XPath.match(document, "//a:EncryptedAssertion", { "a" => ASSERTION })
+
+        unless assertions.length + encrypted_assertions.length == 1
```

Comment:
just a suggestion (personal preference) when using `length` or `size`:
```
Whether you should use length or size is mostly a matter of personal preference. Personally I use size for collections (hashes, arrays, etc) and length for strings, since for me objects like hashes and stacks don’t have a length, but a size (defined in terms of the elements they contain)
```

http://batsov.com/articles/2014/02/17/the-elements-of-style-in-ruby-number-13-length-vs-size-vs-count/